velox/exec/StreamingWindow.cpp

```diff
@@ -0,0 +1,199 @@
+  frameEndBuffers_.reserve(numFuncs);
+  validFrames_.reserve(numFuncs);
+
+  for (auto i = 0; i < numFuncs; i++) {
+    BufferPtr frameStartBuffer = AlignedBuffer::allocate<vector_size_t>(
+        numRowsPerOutput_, operatorCtx_->pool());
+    BufferPtr frameEndBuffer = AlignedBuffer::allocate<vector_size_t>(
+        numRowsPerOutput_, operatorCtx_->pool());
+    frameStartBuffers_.push_back(frameStartBuffer);
+    frameEndBuffers_.push_back(frameEndBuffer);
+    validFrames_.push_back(SelectivityVector(numRowsPerOutput_));
+  }
+}
+
+void StreamingWindow::noMoreInput() {
+  Operator::noMoreInput();
+}
+
+bool StreamingWindow::isFinished() {
+  return noMoreInput_ && data_->numRows() == 0 && output_ == nullptr;
+}
+
+RowVectorPtr StreamingWindow::createFinalResult() {
+  if (output_->size() > outputBatchSize_) {
+    auto result = std::dynamic_pointer_cast<RowVector>(
+        output_->slice(0, outputBatchSize_));
+    auto remainingSize = output_->size() - outputBatchSize_;
+    output_ = std::dynamic_pointer_cast<RowVector>(
+        output_->slice(outputBatchSize_, remainingSize));
+    return result;
+  } else if (noMoreInput_) {
+    auto result = std::move(output_);
+    return result;
+  } else {
+    return nullptr;
+  }
+}
+
+RowVectorPtr StreamingWindow::createOutput() {
+  numRows_ = sortedRows_.size();
+
+  currentPartition_ = 0;
+  numProcessedRows_ = 0;
+
+  createPeerAndFrameBuffers();
+
+  auto result = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(outputType_, numRows_, operatorCtx_->pool()));
+
+  // Set all passthrough input columns.
+  for (int i = 0; i < numInputColumns_; ++i) {
+    data_->extractColumn(sortedRows_.data(), numRows_, i, result->childAt(i));
+  }
+
+  // Construct vectors for the window function output columns.
+  std::vector<VectorPtr> windowOutputs;
+  windowOutputs.reserve(windowFunctions_.size());
+  for (int i = numInputColumns_; i < outputType_->size(); i++) {
+    auto output = BaseVector::create(
+        outputType_->childAt(i), numRows_, operatorCtx_->pool());
+    windowOutputs.emplace_back(std::move(output));
+  }
+
+  // Compute the output values of window functions.
+  callApplyLoop(numRows_, windowOutputs);
+
+  for (int j = numInputColumns_; j < outputType_->size(); j++) {
+    result->childAt(j) = windowOutputs[j - numInputColumns_];
+  }
+
+  // clear the data_, sortedRows_ and partitionStartRows_.
+  data_->eraseRows(
+      folly::Range<char**>(sortedRows_.data(), sortedRows_.size()));
+  sortedRows_.clear();
+  partitionStartRows_.clear();
+
+  if (output_ == nullptr) {
+    output_ = result;
+  } else {
+    output_->append(result.get());
+  }
+  return createFinalResult();
+}
+
+RowVectorPtr StreamingWindow::getOutput() {
+  if (noMoreInput_) {
+    // Handle the last group in last input batch
+    updateValues();
+    partitionStartRows_.push_back(sortedRows_.size());
+  }
+
+  if (sortedRows_.size() > 0) {
+    return createOutput();
```

Comment:
Nit : Make this and the below function names consistent "createOutput" and "createFinalOutput" or "createResult" and "createFinalResult"

Reply:
Updated.